### PR TITLE
[PHASE IV] test(core): add regression tests for use_gpu device visibility

### LINKED ISSUE #765: [PHASE IV] use_gpu: false in testenv.yaml is a no-op, and use_gpu: true is applied after the base model is already constructed
**What happened**:

`use_gpu` is declared in `core/testenvmanager/testenv/testenv.py` and parsed from `testenv.yaml`, but it is read in exactly one place in the core (the matches under `examples/…/GAN/lpips/` are an unrelated constructor argument in a vendored third-party library):

```python
# core/testcasecontroller/algorithm/paradigm/singletask_learning/singletask_learning.py:55-56
if kwargs.get("use_gpu", True):
    os.environ["CUDA_VISIBLE_DEVICES"] = "0"
```

Two separate defects follow from those two lines.

**1. `use_gpu: false` does nothing at all.** There is no `else` branch, so a run configured for CPU never restricts device visibility — it silently inherits whatever `CUDA_VISIBLE_DEVICES` the host has, and the model takes the GPU anyway. `examples/llm-edge-benchmark-suite/single_task_bench_with_compression/testenv/testenv.yaml` ships with `use_gpu: false` and does not get the CPU-only run it asks for.

**2. Even `use_gpu: true` is applied too late to have its intended effect.** Line 50 of the same file calls `ParadigmBase.__init__`, which calls `_get_module_instances()` — and that imports *and instantiates* the user's `basemodel.py`. That is where TensorFlow/PyTorch get imported and where many models select their device. The environment variable is only assigned on line 56, six lines after the model already exists.

Additionally, `use_gpu` is honoured only by the `singletasklearning` paradigm. Every other paradigm ignores the setting entirely, even though it is declared on the shared `TestEnv`.

**What you expected to happen**:

- `use_gpu: false` restricts the run to CPU.
- `use_gpu: true` takes effect before any algorithm module is imported.
- The setting behaves consistently across paradigms, since it lives on the shared `TestEnv`.

**How to reproduce it (as minimally and precisely as possible)**:

Any `singletasklearning` job works. Point a testenv at a base model that just reports what it sees:

```python
# testalgorithms/basemodel.py
import os
from sedna.common.class_factory import ClassType, ClassFactory

__all__ = ["ProbeModel"]

def _report(stage):
    print(f"[PROBE] {stage}: CUDA_VISIBLE_DEVICES={os.environ.get('CUDA_VISIBLE_DEVICES', '__unset__')!r}")

@ClassFactory.register(ClassType.GENERAL, alias="ProbeModel")
class ProbeModel:
    def __init__(self, **kwargs):
        _report("basemodel.__init__")

    def train(self, train_data, valid_data=None, **kwargs):
        _report("train")
        return None

    def save(self, model_path):
        os.makedirs(model_path, exist_ok=True)
        path = os.path.join(model_path, "model.txt")
        with open(path, "w", encoding="utf-8") as f:
            f.write("probe")
        return path

    def load(self, model_url=None):
        return model_url

    def predict(self, data, **kwargs):
        _report("predict")
        return [0 for _ in data]
```

Set an ambient value standing in for whatever devices the host exposes, and run the job. No actual GPU is needed — the probe only reads the variable, so the defect is observable on any machine:

```
export CUDA_VISIBLE_DEVICES=0,1,2,3
ianvs -f ./benchmarkingjob.yaml
```

Observed, on current `main`:

| stage | `use_gpu: true` | `use_gpu: false` |
|---|---|---|
| `basemodel.__init__` | `'0,1,2,3'` — not applied yet | `'0,1,2,3'` |
| `train` | `'0'` | `'0,1,2,3'` — never applied |
| `predict` | `'0'` | `'0,1,2,3'` |

The right-hand column should read `-1` (or equivalent) at every stage. The left-hand column should read `0` at every stage, including `__init__`.

**Anything else we need to know?**:

The fix needs care around one trap. `TestEnv` declares `use_gpu = False` as its default, but the *effective* behaviour when a testenv omits the key has always been "do not touch `CUDA_VISIBLE_DEVICES`" — because the only consumer's fallback is `kwargs.get("use_gpu", True)`, defaulting to `True`. The declared default and the real behaviour contradict each other.

Only 2 of the 41 `testenv.yaml` files under `examples/` declare `use_gpu` at all. So honouring the declared `False` default literally would force every other example onto CPU — a silent, repo-wide regression. Any fix should treat "not configured" as a third state distinct from an explicit `false`.

I have a patch for this and will open a PR shortly.

### PR DESCRIPTION
**Pull request by vjuhhii**
**What type of PR is this?**

/kind test

**What this PR does / why we need it**:

Adds regression tests for the `use_gpu` device-visibility behaviour introduced in #767. Reviewers
on that PR asked for a test covering the `use_gpu: false` path on a multi-GPU host; these are those
tests, kept in a separate PR so #767 stays at +21/−3.

> **Depends on #767.** These tests assert the behaviour that #767 introduces, so they should land
> after it. On current `main` five of them fail by design — see the verification table below.

These are the first tests in the repository. `tests/` is the location `setup.py` already
anticipates:

```python
packages=find_packages(exclude=["tests", "*.tests", "*.tests.*", "tests.*"])
```

The exclusion has been there all along with nothing to exclude, so this takes the place the project
had already reserved rather than inventing a new convention.

### Why these particular tests

#765 reported two defects, and both are ordering-sensitive:

1. `use_gpu: false` was a silent no-op, so a run configured for CPU inherited whatever devices the
   host exposed.
2. `use_gpu: true` was applied *after* the algorithm module had been instantiated, so AI frameworks
   had already read `CUDA_VISIBLE_DEVICES` and fixed their device list before the restriction took
   effect.

A test that checks the environment variable *after* construction finishes would pass against the
original buggy code and prove nothing. So each case here asserts the value of `CUDA_VISIBLE_DEVICES`
**at the moment the algorithm module is instantiated**, using a stand-in module that records it
inside `get_module_instance()` — the point where, in a real run, `basemodel.py` is imported and a
framework picks its device.

Every case starts from `CUDA_VISIBLE_DEVICES=0,1,2,3` to represent the multi-GPU host that made the
original no-op visible.

| Test | Asserts |
|---|---|
| `test_use_gpu_false_hides_every_device_on_a_multi_gpu_host` | `-1` — the CPU-only run does not inherit host devices |
| `test_use_gpu_true_restricts_to_one_device_before_modules_load` | `0` — applied before the module loads, not after |
| `test_unset_use_gpu_leaves_an_existing_value_alone` | `0,1,2,3` — unchanged |
| `test_unset_use_gpu_does_not_introduce_the_variable` | the variable is not created where none existed |
| `test_set_gpu_visibility_maps_each_state` | the three states, independent of paradigm construction |

The two `unset` cases are invariants rather than regressions: 39 of the 41 `testenv.yaml` files
under `examples/` omit `use_gpu`, and their behaviour must not change.

### Verification

| Branch | Result |
|---|---|
| `main` (without #767) | **5 failed, 2 passed** |
| with #767 applied | **7 passed** |

The two that pass on `main` are the invariants above — correct, since #767 deliberately preserves
that path. The five failures are the three defects plus the two cases that reach
`_set_gpu_visibility`, which does not exist before #767.

### Running them

No new runtime dependencies; the suite needs only `pytest` on top of the environment CI already
builds:

```bash
pip install resources/third_party/*
pip install -r requirements.txt
pip install pytest
pytest tests/
```

### Relationship to #845

#845 proposes building a pytest suite for `core/` as Phase IV work — `pytest.ini`, a shared
`conftest.py`, CI integration, and unit tests across all paradigms.

This PR is deliberately not that. It adds the specific regression tests reviewers asked for on
#767, in the location `setup.py` already reserves, and adds no test infrastructure, no manifest
entry for pytest, and no workflow changes. If #845 lands a broader structure, these tests should
move under it — I would rather have the first cases in place and adapt them than block on the
framework.

### Notes

- `pylint` in CI analyses `core/` only, so these files are outside its scope.
- Wiring `pytest tests/` into a workflow is the natural follow-up. I left it out to keep this PR to
  one thing, and because #845 proposes owning that piece.

**Which issue(s) this PR fixes**:

None — this PR closes nothing on its own. It adds test coverage for the behaviour fixed by #767,
which is the PR that resolves #765. Related: #765, #767, #845.

### CODE REVIEW

**Comment by rahulnr107** [comment]
## Review — PR #890

### Summary

PR #890 adds regression tests for the `use_gpu` behavior addressed by PR #767 and Issue #765.

The tests are implemented in:

```text
tests/core/test_paradigm_base.py
```

### Test Coverage

The PR covers the three states of the configuration:

```text
use_gpu = False
use_gpu = True
use_gpu = None
```

It verifies that:

```text
False → CUDA_VISIBLE_DEVICES = "-1"
True  → CUDA_VISIBLE_DEVICES = "0"
None  → existing environment remains unchanged
```

The tests also cover the case where the variable was previously absent.

### Strong Point

The most important aspect of the tests is that they record `CUDA_VISIBLE_DEVICES` at the moment the algorithm module is instantiated.

This is important because the original defect was ordering-sensitive. A test that checked the environment only after construction could pass even if the implementation still configured GPU visibility too late.

The `RecordingModule` approach therefore tests the relevant behavior at the correct boundary without requiring a real GPU.

### Assessment

The tests provide meaningful regression coverage for the Core behavior introduced by PR #767.

### Recommendation

**Accepted as it is.**

### Verification Boundary

The tests use a controlled `RecordingModule` rather than a real AI framework or physical GPU. Therefore they verify the Core ordering and environment-variable contract, but they do not by themselves establish end-to-end GPU behavior with every supported framework.

A separate integration test could provide that additional confidence if required.

### Relationship to PR #767

PR #767 provides the implementation change.

PR #890 provides regression protection for that behavior.

The two PRs are therefore complementary rather than duplicated.

**Comment by vjuhhii** [comment]
Thanks for reviewing, and for stating the verification boundary explicitly — that is exactly the right way to read these. They establish the Core ordering and environment contract; they do not substitute for running a real framework on real hardware.

An integration test would need a GPU runner, which the project does not have today. #888 is my attempt at the next best thing: measure the hardware-dependent behaviour directly and report it, rather than assume it.

**Comment by Hiteshsai007** [comment]
These tests are structured exactly right: asserting `CUDA_VISIBLE_DEVICES` **at the moment the
algorithm module is instantiated** (rather than after construction finishes) is what makes the
ordering-sensitive cases meaningful, and the per-test `monkeypatch` isolation is clean.

Two gaps are worth covering — both involve **multiple constructions in one process**, which is how
ianvs actually runs (one paradigm per TestCase, sequentially):

1. **Cross-TestCase carry-over.** Every case here builds a single `ParadigmBase`. On `pull/767`,
   constructing with `use_gpu=False` sets `CUDA_VISIBLE_DEVICES="-1"` and never restores it, so the
   *next* construction that omits `use_gpu` instantiates its module with `"-1"` still in the
   environment rather than the host value. I traced this with the real `SingleTaskLearning`
   (recording the env at module load across two constructions): after a `use_gpu=False` case, an
   omitted-`use_gpu` case sees `"-1"`. A test shaped like this would lock it down:

   ```python
   monkeypatch.setenv("CUDA_VISIBLE_DEVICES", MULTI_GPU_HOST)
   build_paradigm(tmp_path, use_gpu=False)          # sets "-1", never restored on #767
   assert build_paradigm(tmp_path) == MULTI_GPU_HOST  # currently FAILS (sees "-1")
   ```

2. **Exception path.** If `_get_module_instances()` raises after the visibility value is set,
   there is no `finally`/restore, so the mask persists into the next case. A test that makes the
   module raise mid-construction and then asserts the prior environment is restored would cover it.

These failures are #767's behaviour rather than this PR's, but this test file is the natural place
to pin the **whole** contract (single + sequential + exception) once #767 restores state — otherwise
the regression suite goes green while the two real multi-TestCase paths stay open.

**Comment by vjuhhii** [comment]
Thanks — I looked into both, and they land differently.

Sequential construction: added, though the reachable version is narrower than described. The scenario of one case setting -1 and a later case omitting use_gpu cannot arise through ianvs configuration: TestCaseController.build_testcases() passes the same test_env object to every TestCase it appends, so all cases in a run share one use_gpu value. They cannot disagree about it.

What does matter across constructions is that each one applies the setting itself rather than trusting what the previous one left behind, and that was genuinely untested. Added in 1295573: one case asserts the value is re-applied after something else has restored the host's devices in between, and one covers the gpu path over repeated construction.

Environment restoration on failure: real, but currently unreachable. There is no try/finally around the visibility change, so a failure mid-construction does leave the mask in place. But run_testcases() raises RuntimeError on the first failing case, which aborts the job — there is no later case to be affected. It would become a genuine defect the moment ianvs grows a continue-on-failure mode, so it is worth having on the record; I would rather not add cleanup code for a path that cannot be exercised today, and would defer to maintainers if they see it differently.

**Comment by vjuhhii** [comment]
Correction to something I claimed in the description: I wrote that these are the
first tests in the repository. That is true of `main`, but #648 has been open
since 4 August and also creates `tests/`, with `tests/test_metrics.py`. It got
there first and I should have found it before writing that line.

No file conflict — different filenames — but the two take different approaches to
imports. #648 does `sys.path.insert(0, ...)` at the top of the test module, while
these run from the repository root without it. If both land, that is worth
reconciling, and I am happy to adopt whichever convention maintainers prefer or
to rebase onto #648 once it merges.

Incidentally, #648 also fixes the unbalanced return in `compute()` that made
`bwt_func` raise `ValueError` instead of returning NaN. I ran into that
separately while reading the lifelong learning paradigm; good to see it already
covered there.
